### ISSUE TITLE
[MUSIQ] Will you also release the training code?

### ISSUE THREAD

**Issue by TimHunt12** [google-research/google-research#914]
I am impressed with the approach and work done with MUSIQ for image quality assessment.  I would like to use it to expand the capabilities and even train with additional data for specific IQA instances.

At the moment it appears that this code is for doing a prediction using one of the model checkpoints created from the competitive datasets.  I would like to train the approach on some unique datasets and train it further.

Although I think I may be able to come up with the code to train a model based on the paper, I wanted to see if you would be willing to release your training scripts as well so I could train new datasets quicker.  Thanks.

**Comment by TimHunt12** [comment]
@junjiek 
I understand if you can't share the script that you used to train the model.  I have been studying the paper and the code and feel I can perhaps now come up with a script to do training.  However, I have a question. I suspect that to train the model I would need to provide the images and our MOS scores.  However, I am not sure what format they need to be in. Do the images need to be .jpg?  How about the MOS scores.  What is the best way to provide the model the MOS score for each image?  If you can help I would appreciate it as I would like to see if I can make this model focus on a specific kind of image and use the existing models as a transformer to focus the MOS scores for these images.    Thanks.

**Comment by junjiek** [comment]
Hi Tim,

Thanks a lot for your interest in the paper and code. At this point we
don't have plans to open source the training code yet.

For training the model, the image format can be anything. In our code we
use tf.image.decode_jpeg() by default. The MOS is a single value for each
image. For AVA, the label for each image is a score distribution.  In
training, we just feed the (image, MOS) pair to the model.

Hope it helps,
Junjie

**Comment by TimHunt12** [comment]
@junjiek  
Xie Xie Junjie,

I appreciate you getting back to me.  

I saw the use of tf.image.decode_jpeg() in the preprocessor.py and I think I can change it to tf.image.decode_image to handle .png and .bmp along with .jpg or jpeg.  That will save me from prepping images to be in a specific format. Provided that the encoding of those files don't require additional handling of the features in preprocessing.  I haven't tested that yet so I will have to see if I can do that.

I also have another question.  Is there a reason in the preprocessor.py file under the get_hashed_spactial_pos_emb_index function your team repeated this line of code `pos_emb_hash_h = tf.cast(pos_emb_hash_h, dtype=tf.int32)`. I'm not sure I understand the benefit of running the same code here twice.

You mentioned above that you just sent in the (image, MOS) pair to the model for training.  I assume the image was the "image path" to an image based on how the code processes it and the MOS or (mean opinion score) was just a float value between 1 and 5.  To feed in a whole batch of images did you just feed in an array of image path/names and MOS scores in a list of tuples or did you feed each in as an array and zip them together or did you prepare a dictionary file of them to feed to the application?  Sorry rather long question.

Actually I am not even sure my question above is the right one.  How the data is fed into the model is more determined by how I prepare it to train.  I can see points in the code where possibly training type code was potentially removed.  Since you aren't going to release that code I will have to figure out if I can come up with it on my own training code based on the paper and the code you have released.

I have learned so much studying your team's code.  I love your use of Jax to speed up GPU processing of images and I didn't even know until now that Tensorflow had a tf.image.extract_patches() function.  That must the advantage of working with Google.  You either learn quickly about the advances the Google Research team does in expanding Tensorflow or you are able to suggest and maybe help them build them.  Either way it is very nice work.

Please again accept my appreciation for your team's great skill in building such a clean simple model.  It is very impressive.  I may not be able to come up with a way to leverage it and extend it into other fields of use and may have to use another model to meet my short term needs but nice work and I am enjoying studying your code.

Thanks
Tim

**Comment by junjiek** [comment]
Hi Tim,

Thank you for the recognition for the paper and code.

1) pos_emb_hash_h = tf.cast(pos_emb_hash_h, dtype=tf.int32) should just be
called once. I will remove the redundant line. Thanks for pointing it out.

2) We save each (image, MOS) pair as tf records.
https://www.tensorflow.org/tutorials/load_data/tfrecord#reading_a_tfrecord_file.
More specifically, we use the API of tensorflow dataset. For example, this
is the publicly available imagenet dataset :
https://www.tensorflow.org/datasets/catalog/imagenet2012

Thanks again for the kind words. Hope the answers are helpful.

Best,
Junjie

**Comment by TimHunt12** [comment]
@junjiek 

Thank you.  You have been very helpful.  Again I am very impressed with your teams work.  Thanks,

Tim

**Comment by TimHunt12** [comment]
@junjiek

I was taken off of this project for a little while and just got reassigned.  I am trying to run a prediction with the MUSIQ code here using one of the existing checkpoints.  

I discovered quickly that some of the library references were depreciated and I have update the libraries as well as TensorFlow to 2.9.1.  I am trying to get an MOS score on 1 image and when the code gets to line 143 of run_predict_image.py
'model = model_mod.Model.partial(
      num_classes=num_classes, train=False, **model_config)'
 it is trying to call the .partial attribute of the Model class from multiscale_transformer.py and it fails.  When I look in the code for multiscale_transformer.py there is no class function partial.
 
When the prediction code for MUSIQ was posted to the GitHub, was there some code left out that was supposed to be here to run the prediction or am I missing something?

Thanks,   Tim
